tests/kernel/mem_protect/userspace/src/main.c

```diff
@@ -0,0 +1,231 @@
+/*
+ * Parts derived from tests/kernel/fatal/src/main.c, which has the
+ * following copyright and license:
+ *
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <ztest.h>
+#include <kernel_structs.h>
+#include <string.h>
+#include <stdlib.h>
+
+#define INFO(fmt, ...) printk(fmt, ##__VA_ARGS__)
+
+/* ARM is a special case, in that k_thread_abort() does indeed return
+ * instead of calling _Swap() directly. The PendSV exception is queued
+ * and immediately fires upon completing the exception path; the faulting
+ * thread is never run again.
+ */
+#ifndef CONFIG_ARM
+FUNC_NORETURN
+#endif
+void _SysFatalErrorHandler(unsigned int reason, const NANO_ESF *pEsf)
+{
+	INFO("Caught system error -- reason %d\n", reason);
+	ztest_test_pass();
+#ifndef CONFIG_ARM
+	CODE_UNREACHABLE;
+#endif
+}
+
+static void is_usermode(void)
+{
+	/* Confirm that we are in fact running in user mode. */
+	zassert_true(_is_user_context(), "thread left in kernel mode\n");
+}
+
+static void priv_insn(void)
+{
+	/* Try to invoke what should require a privileged instruction. */
+	unsigned int key = irq_lock();
+
+	irq_unlock(key); /* just in case we succeeded */
+	zassert_unreachable("irq_lock did not fault\n");
+}
+
+static void write_control(void)
+{
+	/* Try to write to a control register. */
+#if defined(CONFIG_X86)
+	__asm__ volatile (
+		"mov %cr0, %eax;\n\t"
+		"and $0xfffeffff, %eax;\n\t"
+		"mov %eax, %cr0;\n\t"
+		);
+#elif defined(CONFIG_ARM)
+	__asm__ volatile (
+		"mrs r2, CONTROL;\n\t"
+		"bic r2, #1;\n\t"
+		"msr CONTROL, r2;\n\t"
+		);
+#endif
+	zassert_unreachable("Write to control register did not fault\n");
+}
+
+static void disable_mmu_mpu(void)
+{
+	/* Try to disable memory protections. */
+#if defined(CONFIG_X86)
+	__asm__ volatile (
+		"mov %cr0, %eax;\n\t"
+		"and $0x7ffeffff, %eax;\n\t"
+		"mov %eax, %cr0;\n\t"
+		);
+#elif defined(CONFIG_ARM)
+	arm_core_mpu_disable();
+#endif
+	zassert_unreachable("Disable MMU/MPU did not fault\n");
+}
+
+static void read_kernram(void)
+{
+	/* Try to read from kernel RAM. */
+	void *p = _current->init_data;
+
+	printk("%p\n", p);
+	zassert_unreachable("Read from kernel RAM did not fault\n");
+}
+
+static void write_kernram(void)
+{
+	/* Try to write to kernel RAM. */
+	_current->init_data = NULL;
+	zassert_unreachable("Write to kernel RAM did not fault\n");
+}
+
+extern int _k_neg_eagain;
+
+static void write_kernro(void)
+{
+	/* Try to write to kernel RO. */
+	_k_neg_eagain = -EINVAL;
```

Comment:
I'm wondering if there's a say to static_assert() that a symbol is indeed in a particular section.

Reply:
Don't know how to do this, but willing to do so if someone else does.

Reply:
Played around with this a bit, but couldn't make it work.  Expression was non-constant.  Also tried making it a runtime assertion via zassert_true() in the test, and that worked on x86 but didn't build on arm (_image_rodata_start/end aren't defined there).

Reply:
You could check if &_k_neg_eagain falls within the bounds of __kernel_ram_start and __kernel_ram_end since it should be in the read/write data section for the kernel.

Reply:
k_neg_eagain is actually in rodata (it is actually const; I intentionally mis-declared it here in the extern so I could write to it without compiler complaint, although alternatively I could use the correct decl and use pointer aliasing as I did in the protection test for writing rodata). But the real problem here is that if I add a BUILD_ASSERT that attempts to compare with either _image_rodata_start or __kernel_ram_start, the compiler says error: expression in static assertion is not constant.

Reply:
from my perspective doing the check at runtime should be fine.
If ARM's linker script doesn't declare some of the special symbols in include/linker/linker-defs.h we should add them.

Reply:
Yes, doing the check at runtime is fine if it's not possible to do it at compile time.  It's just a test, size and performance are not an issue.

Reply:
Request for clarification: Did you want me to still try to add this to this same change, or just leave this one alone and do a new PR to add it later? Wasn't sure since this one has now been approved.

Reply:
I went ahead and added the changes on top of the existing one.